flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+
+It is worth being familiar with `fir::ExtendedValue`. They described an entity (symbol,
+or evaluated expression) with all its associated dynamic properties (length, extents, lower bounds...).
+A `fir::ExtendedValue` is a union of different SSA value containers that describe a particular category of entities.
+The categories are:
+- `fir::Unboxed` for scalar numericals, logicals, and non-polymorphic derived types without length parameters,
+- `fir::CharBoxValue` for scalar characters,
+- `fir::ArrayBoxValue` for contiguous arrays of numerical, logical, and non-polymorphic derived types without length parameters.
+- `fir::CharArrayBoxValue` contiguous character arrays,
+- `fir::MutableBoxValue` for allocatable and pointers of all types.
+- `fir::BoxValue` for all the rest.
+See `include/flang/Support/BoxValue.h` for more details.
+
+If the intrinsic implementation use runtime, the actual binding should be done in another helper function called
+`gen{Runtime function name}`. For instance character runtime bindings are implemented in `lib/Character/CharacterRuntime.cpp`
+See `genTrim` for an example of such runtime binding helper.
+Non character related runtime binding should be added to `lib/Character/Runtime.cpp`.
+
+Note: Numerical elemental intrinsics that have libm/libnvmath implementation are automatically mapped and
+do not follow the above patterns (this include intrinsics like acos, bessel_j...).
+
+#### Tests
+LLVM FileCheck regression tests are used in lowering. See `test/Lower/intrinsic-procedures.f90`.
+To get familiar with FileCheck tests, it is worth having a look at `https://llvm.org/docs/CommandGuide/FileCheck.html`.
+A few test cases must be added in `intrinsic-procedures.f90` to test that the fir output looks like what is expected.
+To do quick eye checking during development, the FIR produced for a Fortran file can be dumped with `bbc -emit-fir`.
+
+### Adding intrinsics in the FIR and Codegen
+So far, nothing to do there. No FIR operation implements Fortran intrinsic procedures.
+
+## Tips
+
+Be sure to respect alphabetical order when adding a new intrinsic in one of the tables or
+if statements. They are usually sorted according to intrinsic names.
+
+## An End-to-End example: TRIM
+TRIM is a transformational intrinsic operating on characters.
+It takes a scalar character and returns a scalar character. See Fortran standard for more details about it.
+### TRIM name resolution and semantic checks
+In lib/Evaluate/intrinsics.cpp, the table contains a line:
+```
+ {"trim", {{"string", SameChar, Rank::scalar}}, SameChar, Rank::scalar, IntrinsicClass::transformationalFunction}
+```
+If defines a transformational intrinsic named "trim" that takes a scalar character argument
+named "string" and returns a scalar character of the same kind as the argument.
+There is nothing more needed for name resolution and semantic checks.
+
+### TRIM front-end folding
```

Comment:
It would be good to have an example of a test for folding TRIM in this section.

Reply:
I added an example of constant folding test for TRIM in in https://github.com/flang-compiler/f18-llvm-project/commit/11b8022d47badec9430f126523630e4043969416